WeartGloves/src/WeArtGlove.cpp

```diff
@@ -0,0 +1,1598 @@
+        {
+            for (int j = 0; j < humanJointState.cols(); j++)
+            {
+                humanJointState(i, j) = std::max(0.0, std::min(1.0, humanJointState(i, j)));
+            }
+        }
+        humanJointState = humanJointState.cwiseProduct((MAX_JOINT_LIMIT - MIN_JOINT_LIMIT)) + MIN_JOINT_LIMIT;
+    }
+    else if (trackingAlgorithmType == "ASTA")
+    {
+        // Set the desired distance for the distance task
+        thumbDistanceTaskptr->setDesiredDistance(weartRawDataTof(0) / 1000.0);  // thumb
+        indexDistanceTaskptr->setDesiredDistance(weartRawDataTof(1) / 1000.0);  // index
+        middleDistanceTaskptr->setDesiredDistance(weartRawDataTof(2) / 1000.0); // middle
+
+        weartRawDataAccRotated = fingerSensorToBodyRotation * weartRawDataAcc.transpose();
+        weartRawDataGyroRotated = fingerSensorToBodyRotation * weartRawDataGyro.transpose();
+
+        // Convert Degree to Radians
+        weartRawDataGyroRotated = weartRawDataGyroRotated * EIGEN_PI / 180.0;
+
+        // Set the desired gravity direction for the gravity task
+        thumbGravityTaskptr->setDesiredGravityDirectionInTargetFrame(weartRawDataAccRotated.col(0));
+        indexGravityTaskptr->setDesiredGravityDirectionInTargetFrame(weartRawDataAccRotated.col(1));
+        middleGravityTaskptr->setDesiredGravityDirectionInTargetFrame(weartRawDataAccRotated.col(2));
+        palmGravityTaskptr->setDesiredGravityDirectionInTargetFrame(palmSensorToBodyRotation * weartRawDataAcc.row(3).transpose());
+
+        const auto &[basePosition, baseRotation, jointPosition] = integrator->getSolution();
+
+        // update the KinDynComputations object
+        baseTransform.topLeftCorner<3, 3>() = baseRotation.rotation();
+        baseTransform.topRightCorner<3, 1>() = basePosition;
+
+        // solve the IK
+        ikProblem->ik->advance();
+
+        // get the output of the IK
+        baseVelocity = ikProblem->ik->getOutput().baseVelocity.coeffs();
+        jointVelocity = ikProblem->ik->getOutput().jointVelocity;
+
+        // propagate the dynamical system
+        dynamics->setControlInput({baseVelocity, jointVelocity});
+        integrator->integrate(0ms, dT);
+
+        kinDyn->setRobotState(baseTransform,
+                              jointPosition,
+                              baseVelocity,
+                              jointVelocity,
+                              gravity);
+
+        if (PalmSidePrefix == "Right")
+        {
+            humanJointState = closureToHumanJoint * jointPosition;
+        }
+        else if (PalmSidePrefix == "Left")
+        {
+            humanJointState = closureToHumanJoint * -jointPosition;
+        }
+    }
+
+    return true;
+}
+
+bool WeArtGlove::WeArtGloveImpl::close()
+{
+    for (auto &[key, element] : thimbleInformationMap)
+    {
+        element.hapticObject->RemoveEffect(element.thimbleEffect.get());
+    }
+    
+    yarp::os::Time::delay(1);
```

Comment:
Why do you need this delay here?

Reply:
Well, this is related to closing of the device properly. I define properly as this: When trying to close the yarp device using `ctrl+c` I am also interested the Middlware also stops. Sometimes (couldn't find the cause yet!) the yarp device closes but the Middleware is still running. I defined this delay to give the middleware sometime to stop. I also need to mention that with their latest SDK they added some callbacks which may fix this issue. I will address them.